providers/adfs.go

```diff
@@ -0,0 +1,81 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"net/url"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid email profile"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	defaultScope := adfsDefaultScope
+
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		resource := p.ProtectedResource.String()
+		if !strings.HasSuffix(resource, "/") {
+			resource += "/"
+		}
+
+		if p.Scope != "" && !strings.HasPrefix(p.Scope, resource) {
+			p.Scope = resource + p.Scope
+		}
+		defaultScope = resource + defaultScope
+	}
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: defaultScope,
+	})
+
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL Override to double encode the state parameter
```

Comment:
Can you elaborate on why this is needed? Potentially with a reference link for why ADFS needs double encoded state?

Reply:
Also, OIDCProvider just had added functionality where the OIDC nonce part of the OAuth flow was added -- take a look at master so you can handle that change.

Reply:
I had to add double encoding cause if not, the get parameters were lost after the redirection. I'm not sure if it's an ADFS bug. I found this: 
https://github.com/iay/mdq-server/issues/38
https://stackoverflow.com/questions/30835670/adfs-relying-party-trust-which-has-a-querystring-parameter

And if you check the docs here: https://docs.microsoft.com/en-us/powerapps/maker/portals/configure/configure-saml2-settings

> The string value is encoded and placed into a container string of the format RPID=&lt;URL encoded RPID&gt;&RelayState=&lt;URL encoded RelayState&gt;. This entire string is once again encoded

Regarding the nonce I will set SkipNonce: true

Reply:
Sounds good! 

We've had another issue raised with a similar problem for IBM Cloud OIDC I think it was.

Can you explain this ADFS bugfix hack in a comment in the code?